File: php/world.php
<?php

error_reporting(E_ALL);

require_once 'cell.php';

final class LocationOccupied extends Exception {
  public function __construct(
    public readonly int $x,
    public readonly int $y,
  ) {
    parent::__construct("LocationOccupied($x-$y)");
  }
}

final class World {
  public int $tick = 0;

  private array $cells = [];

  private const array DIRECTIONS = [
    [-1, 1],  [0, 1],  [1, 1],  // above
    [-1, 0],           [1, 0],  // sides
    [-1, -1], [0, -1], [1, -1], // below
  ];

  public function __construct(
    private readonly int $width,
    private readonly int $height,
  ) {
    $this->populateCells();
    $this->prepopulateNeighbours();
  }

  public function doTick(): void {
    // First determine the action for all cells
    foreach ($this->cells as $cell) {
      $aliveNeighbours = $cell->aliveNeighbours();
      if (!$cell->alive && $aliveNeighbours === 3) {
        $cell->nextState = true;
      } elseif ($aliveNeighbours < 2 || $aliveNeighbours > 3) {
        $cell->nextState = false;
      } else {
        $cell->nextState = $cell->alive;
      }
    }

    // Then execute the determined action for all cells
    foreach ($this->cells as $cell) {
      $cell->alive = $cell->nextState;
    }

    $this->tick += 1;
  }

  public function render(): string {
    // The following is the fastest
    $rendering = '';
    for ($y = 0; $y < $this->height; $y++) {
      for ($x = 0; $x < $this->width; $x++) {
        $cell = $this->cellAt($x, $y);
        if ($cell) {
          $rendering .= $cell->toChar();
        }
      }
      $rendering .= "\n";
    }
    return $rendering;

    // The following is slower
    // $rendering = array();
    // for ($y = 0; $y < $this->height; $y++) {
    //   for ($x = 0; $x < $this->width; $x++) {
    //     $cell = $this->cellAt($x, $y);
    //     if ($cell) {
    //       $rendering[] = $cell->toChar();
    //     }
    //   }
    //   $rendering[] = "\n";
    // }
    // return join($rendering);
  }

  private function makeKey(int $x, int $y): string {
    // The following is the fastest
    return "$x-$y";

    // The following is slower
    // return $x . '-' . $y;

    // The following is slower
    // return implode('-', [$x, $y]);
  }

  private function cellAt(int $x, int $y): ?Cell {
    $key = $this->makeKey($x, $y);
    return $this->cells[$key] ?? null;
  }

  private function populateCells(): void {
    for ($y = 0; $y < $this->height; $y++) {
      for ($x = 0; $x < $this->width; $x++) {
        $alive = rand(0, 99) < 20;
        $this->addCell($x, $y, $alive);
      }
    }
  }

  private function addCell(int $x, int $y, bool $alive = false): bool {
    $existing = $this->cellAt($x, $y);
    if ($existing) {
      throw new LocationOccupied($x, $y);
    }

    $key = $this->makeKey($x, $y);
    $cell = new Cell($x, $y, $alive);
    $this->cells[$key] = $cell;
    return true;
  }

  private function prepopulateNeighbours(): void {
    foreach ($this->cells as $cell) {
      $x = $cell->x;
      $y = $cell->y;

      foreach (self::DIRECTIONS as [$relX, $relY]) {
        $nx = $x + $relX;
        $ny = $y + $relY;
        if ($nx < 0 || $ny < 0) {
          continue; // Out of bounds
        }

        if ($nx >= $this->width || $ny >= $this->height) {
          continue; // Out of bounds
        }

        $neighbour = $this->cellAt($nx, $ny);
        if ($neighbour) {
          $cell->neighbours[] = $neighbour;
        }
      }
    }
  }
}

File: ruby/world.rb
require_relative 'cell'
require 'stringio'

class World
  attr_reader :tick

  class LocationOccupied < Exception
    def initialize(x, y)
      super("LocationOccupied(#{x}-#{y})")
    end
  end

  DIRECTIONS = [
    [-1, 1],  [0, 1],  [1, 1], # above
    [-1, 0],           [1, 0], # sides
    [-1, -1], [0, -1], [1, -1] # below
  ]

  def initialize(width:, height:)
    @tick = 0
    @width = width
    @height = height
    @cells = {}

    populate_cells
    prepopulate_neighbours
  end

  def dotick
    # First determine the action for all cells
    @cells.each_value do |cell|
      alive_neighbours = cell.alive_neighbours
      if !cell.alive && alive_neighbours == 3
        cell.next_state = true
      elsif alive_neighbours < 2 || alive_neighbours > 3
        cell.next_state = false
      else
        cell.next_state = cell.alive
      end
    end

    # Then execute the determined action for all cells
    @cells.each_value do |cell|
      cell.alive = cell.next_state
    end

    @tick += 1
  end

  def render
    # The following is slower
    # rendering = ""
    # @height.times.each { |y|
    #   @width.times.each { |x|
    #     cell = cell_at(x, y)
    #     if cell
    #       rendering << cell.to_char
    #     end
    #   }
    #   rendering << "\n"
    # }
    # rendering

    # The following is the fastest
    rendering = []
    @height.times.each { |y|
      @width.times.each { |x|
        cell = cell_at(x, y)
        if cell
          rendering << cell.to_char
        end
      }
      rendering << "\n"
    }
    rendering.join

    # The following is slower
    # rendering = StringIO.new
    # @height.times.each { |y|
    #   @width.times.each { |x|
    #     cell = cell_at(x, y)
    #     if cell
    #       rendering << cell.to_char
    #     end
    #   }
    #   rendering << "\n"
    # }
    # rendering.string
  end

  private

  def make_key(x, y)
    # The following is the fastest
    "#{x}-#{y}"

    # The following is slower
    # x.to_s + '-' + y.to_s

    # The following is slower
    # [x, y].join('-')
  end

  def cell_at(x, y)
    key = make_key(x, y)
    @cells[key]
  end

  def populate_cells
    @height.times do |y|
      @width.times do |x|
        alive = (rand <= 0.2)
        add_cell(x, y, alive)
      end
    end
  end

  def add_cell(x, y, alive = false)
    existing = cell_at(x, y)
    if existing
      raise LocationOccupied.new(x, y)
    end

    key = make_key(x, y)
    cell = Cell.new(x, y, alive)
    @cells[key] = cell
    true
  end

  def prepopulate_neighbours
    @cells.each_value do |cell|
      x = cell.x
      y = cell.y

      cell.neighbours =
        DIRECTIONS.filter_map { |rel_x, rel_y|
          nx = x + rel_x
          ny = y + rel_y
          if nx < 0 || ny < 0
            next # Out of bounds
          end

          if nx >= @width || ny >= @height
            next # Out of bounds
          end

          cell_at(nx, ny)
        }
    end
  end
end
